### ISSUE TITLE
Flickering Pong POMDP

### ISSUE THREAD

**Issue by N-Kingsley** [qfettes/DeepRL-Tutorials#4]
Hi,

Do you know how to convert 'Pong' to 'Flickering Pong POMDP' in original paper?

Thanks,
Ni

**Comment by qfettes** [comment]
Sure. Flickering pong just means that with some probability p, the agent receives a blank screen as the observation rather than the actual screen. 

so after the line
`observation, reward, done, _ = env.step(action)`

add
```
p = 0.5
prob_flicker = random.uniform(0, 1)
if prob_flicker > p:
    observation = np.zeros_like(observation)
```

**Comment by N-Kingsley** [comment]
In this case, can the 'DRQN' also perform well? Is this probability 0.5 too big?

**Comment by qfettes** [comment]
From the paper

> On the Flickering Pong domain, performance
> scales with the observability of the domain, reaching near perfect
> levels when every game screen is observed. This result
> indicates that the recurrent network learns policies that
> are both robust enough to handle to missing game screens
> and scalable enough to improve performance as observability
> increases. Generalization also occurs in the opposite direction:
> when trained on standard Atari games and evaluated
> against flickering games, DRQN’s performance generalizes
> better than DQN’s at all levels of partial information

So, if by well you mean "as good as it would without flickering (or p=0)" then no. However, it does handle partial observability better than DQNs; by this, the authors mean that the performance of DRQN degrades less than DQN as p increases. Figure 5 shows performance of training on normal pong and testing on flickering pong for DQN and DRQN

**Comment by N-Kingsley** [comment]
Thank you for your answer. 

In DRQN.ipynb, What does the abscissa of the reward curve stand for? episode? Does it means 1500000 frames just run for 600 episodes?

**Comment by N-Kingsley** [comment]
I would like to ask how many frames at least needed to have obvious training effect. Why is it still not effective after I trained 1000000 frames?

**Comment by qfettes** [comment]
The x axis would represent the number of completed episodes. I'll make a note and label the axes better. 

I'm looking into the second issue to see if I can reproduce your result. You should see a significant increase in performance by 1 million frames.

**Comment by N-Kingsley** [comment]
Why is it still not effective after I trained 1000000 frames?

I reproduced your code with tensorflow, but not used two tricks that Masking the first half of the losses and gradient truncation. and when I trained it after 1000000 frames, it still score only -21 point. Why?

**Comment by qfettes** [comment]
If it's in code that you reproduced yourself, it's possible that you introduced a bug. Otherwise, small differences between the libraries can lead to large changes in performance. The hyperparameters used here helped train fast on pong, but are not the hyperparameters used in the original publication

**Comment by N-Kingsley** [comment]
I used the hyperparameters in your code, but reproduced it in tensorflow. What should I do?

**Comment by qfettes** [comment]
If it's in code that you wrote, then I'd recommend trying to reproduce the original paper by modifying the hyperparameters, loss function, and optimizer, then checking for bugs. Since it doesn't directly pertain to this repository, I'll be closing this issue soon.

**Comment by N-Kingsley** [comment]
Ok, thank you for your patience

**Comment by N-Kingsley** [comment]
Excuse me, I have another question about DRQN. I run the code for 500000 frames, and I saw a significant increase. When I change 'config.SEQUENCE_LENGTH' from 8 to 4, it can't have a increase after 1500000 frames. This can be understood as the amount of information provided by 4 frames is not enough. But the same result happened when config.SEQUENCE_LENGTH=10 after 500000 frames. Why?

In theory, 10 frames provide more information than 8 frames, and it can also produce good results, although it may take longer.

**Comment by qfettes** [comment]
@N-Kingsley I suspect this is an issue of hyperparameter tuning, but it's hard to say from your message alone. It can be harder for the recurrent network to "make sense" of these longer sequences as well; changing the sequence length from 8 to 10 or 8 to 4 could require you to change several other hyperparameters to learn effectively.
